core/peerset/src/slots.rs

```diff
@@ -62,39 +83,36 @@ impl Slots {
 		let max_slots = max_slots as usize;
 		Slots {
 			max_slots,
-			slots: LinkedHashMap::with_capacity(max_slots),
+			reserved: LinkedHashMap::new(),
+			common: LinkedHashMap::new(),
 		}
 	}
 
 	/// Returns true if one of the slots contains given peer.
 	pub fn contains(&self, peer_id: &PeerId) -> bool {
-		self.slots.contains_key(peer_id)
+		self.common.contains_key(peer_id) || self.reserved.contains_key(peer_id)
 	}
 
 	/// Tries to find a slot for a given peer and returns `SlotState`.
 	pub fn add_peer(&mut self, peer_id: PeerId, slot_type: SlotType) -> SlotState {
-		if let Some(st) = self.slots.get_mut(&peer_id) {
-			if *st == SlotType::Common && slot_type == SlotType::Reserved {
-				*st = SlotType::Reserved;
+		if self.reserved.contains_key(&peer_id) {
+			return SlotState::AlreadyConnected(peer_id);
```

Comment:
i notice that it doesn't do a corresponding  "downgrade" of the slot type, which means that there is no invariant at the end of the function that `peer_id` is in a slot of `slot_type`. Can it have any fallout? I would lean towards documenting those behaviors